### ISSUE TITLE
client-go doesn't properly handle reloading trust anchors during cluster CA rotation

### ISSUE THREAD

**Issue by ahmedtd** [kubernetes/kubernetes#119483]
### What happened?

Discussed in SIG Auth: https://docs.google.com/document/d/1woLGRoONE3EBVx-wTb4pvp4CI7tmLZ6lS26VTbosLKM/edit#bookmark=id.5mfr9vowfkwr

The kube-apiserver serving root CA certificate is injected into pods at /var/run/secrets/kubernetes.io/serviceaccount/ca.crt.

Client-go reads it once [here](https://github.com/kubernetes/client-go/blob/6c596fdbc11ce7779ee320c6a11df44c5d516d54/rest/config.go#L528).  Normally the transport is supposed to automatically reload the roots, but it doesn't set reloadTLSFiles if a CAFile is set [here](https://github.com/kubernetes/kubernetes/blob/928adb27abf755b402be7594844e3c248b0b7670/staging/src/k8s.io/client-go/transport/transport.go#L155-L168).

The net effect is that most clients using InClusterConfig will find their connection to kube-apiserver permanently broken during a rotation of the kube-apiserver serving CA, and won't recover until they are restarted.

We should:
* Set reloadTLSFiles when CAFile is set.
* Make sure that reloadTLSFiles will actually re-read the CAFile.
* Add a test.

### What did you expect to happen?

N/A

### How can we reproduce it (as minimally and precisely as possible)?

N/A

### Kubernetes version

N/A

### Cloud provider

N/A

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by p0lyn0mial** [comment]
@carlory hey, are you working on addressing this issue ?

**Comment by carlory** [comment]
@p0lyn0mial If you want to pick up it, please go ahead.

**Comment by stlaz** [comment]
@p0lyn0mial feel free to take, or I'd have a go later :slightly_smiling_face:

**Comment by p0lyn0mial** [comment]
okay, thanks guys!

/assign @p0lyn0mial

**Comment by tonyaw** [comment]
@p0lyn0mial , may I ask when the fix will be released?
Currently, our product met "certificate signed by unknown authority" error during kube-apiserver root ca.crt renewal, and connection to kube-apiserver can't recover until pod restart.

**Comment by p0lyn0mial** [comment]
@ahmedtd I think that the suggested solution of setting `reloadTLSFiles` and ensuring that the new CA file will be re-read won't work. Please feel free to correct me. 

The reason for this is that the standard library's `tls.Config` currently lacks an exposed callback that would allow us to dynamically inject the new CA, similar to how `GetClientCertificate` is used to dynamically inject a client certificate. Modifying the `tls.Config` is discouraged and [not thread-safe](https://github.com/kubernetes/kubernetes/blob/7e7f1fbc1073c78c9215ef73e25c74123fb2ca3c/staging/src/k8s.io/client-go/rest/rootca_reload_test.go#L151). There have been previous unsuccessful attempts to add such a callback, as seen in this issue: https://github.com/golang/go/issues/22836#issuecomment-576898073.

It seems that we have at least the following options:
- Use `VerifyPeerCertificate` to verify a server certificate. The implementation could be similar to the one used for `GetClientCertificate`. I have created [a simple unit test to demonstrate](https://github.com/kubernetes/kubernetes/blob/7e7f1fbc1073c78c9215ef73e25c74123fb2ca3c/staging/src/k8s.io/client-go/rest/rootca_reload_test.go#L195) how it could work. The drawback of this approach is that we would have to **disable** automatic certificate and host verification by setting InsecureSkipVerify to true.
- Create a custom TLS dialer that allows for thread-safe swapping of the tls.Config. I'll explore what it would take
- Add a callback function to the `tls.Config` to enable the reloading or swapping of the `RootCAs`. that would be my preferable option.

**Comment by stlaz** [comment]
An alternative solution:
You would return a pointer to shallow copies of `tls.Config` with updated RootCAs during `tls.Config.GetConfigForClient`.

**Comment by p0lyn0mial** [comment]
Given my [previous comment](https://github.com/kubernetes/kubernetes/issues/119483#issuecomment-1798488277), it looks like the only viable option right now would be to use `VerifyPeerCertificate`. Are we okay with that?

**Comment by enj** [comment]
> Given my [previous comment](https://github.com/kubernetes/kubernetes/issues/119483#issuecomment-1798488277), it looks like the only viable option right now would be to use `VerifyPeerCertificate`. Are we okay with that?

I am generally against any approach that sets `InsecureSkipVerify`.  Can we close the TLS connection when we the CA file changes?

**Comment by enj** [comment]
We may need to (ab)use something at the round tripper layer to make this work 🤔

**Comment by aojea** [comment]
> > We may need to (ab)use something at the round tripper layer to make this work 🤔
> 
> As far as I know, there isn't anything we could use at the RoundTripper level. We could inject a custom TLS dialer but that [breaks when a proxy is requested](https://github.com/kubernetes/kubernetes/issues/119483#issuecomment-1798488277).

We are already doing that, don't we?

https://github.com/kubernetes/kubernetes/blob/735804dc812ce647f8c130dced45b5ba4079b76e/staging/src/k8s.io/client-go/transport/cache.go#L116-L123

**Comment by p0lyn0mial** [comment]
@aojea this is different, this is creating and assigning a TCP Dialer (`transport.DialContext`). In order to hot reload the `tls.Config.RootCA` we need a TLS Dialer (`transport.DialTLSContext`).

**Comment by aojea** [comment]
> @aojea this is different, this is creating and assigning a TCP Dialer (`transport.DialContext`). In order to hot reload the `tls.Config.RootCA` we need a TLS Dialer (`transport.DialTLSContext`).

I see, but can't we modify that certdialer to rotate the root CA too?

https://github.com/kubernetes/kubernetes/blob/fc691e8ffdb670d5e26a99fc107f1f848e6aa74e/staging/src/k8s.io/client-go/transport/cert_rotation.go#L64

**Comment by p0lyn0mial** [comment]
> I see, but can't we modify that certdialer to rotate the root CA too?

no, because the standard library's `tls.Config` currently lacks an exposed callback that would allow us to dynamically inject the new CA. You will find more info at https://github.com/kubernetes/kubernetes/issues/119483#issuecomment-1798488277

**Comment by enj** [comment]
> can't we modify that certdialer to rotate the root CA too?

No because the TLS config doesn't have a hook for it.  That code works by relying on the TLS config hook that can replace the client cert used for the connection.  No good hook exists to handle the CA changing.  I think we could emulate such a hook by atomically swapping the entire transport at round trip time, but this would only work for users of the round tripper.

**Comment by p0lyn0mial** [comment]
> I think we could emulate such a hook by atomically swapping the entire transport at round trip time, but this would only work for users of the round tripper.

I'm not sure if this would be possible, feel free to post more details :)

**Comment by enj** [comment]
@p0lyn0mial see #122749 (I didn't test it but it should work)

**Comment by p0lyn0mial** [comment]
> @p0lyn0mial see https://github.com/kubernetes/kubernetes/pull/122749 (I didn't test it but it should work)

@enj awesome! I think it can actually work! Thanks, will you have time to finish the PR?

**Comment by enj** [comment]
> > @p0lyn0mial see #122749 (I didn't test it but it should work)
> 
> @enj awesome! I think it can actually work! Thanks, will you have time to finish the PR?

@p0lyn0mial feel free to take over the PR.  I am a bit over committed atm :)

**Comment by p0lyn0mial** [comment]
> @p0lyn0mial feel free to take over the PR. I am a bit over committed atm :)

@tkashem would you be interested in working on the fix for client-go ? Given my current commitments, I'm not sure if I will be able to fix it this release.

**Comment by tkashem** [comment]
> > @p0lyn0mial feel free to take over the PR. I am a bit over committed atm :)
> 
> @tkashem would you be interested in working on the fix for client-go ? Given my current commitments, I'm not sure if I will be able to fix it this release.

sure, i am happy to work on it, thanks! 
/assign

**Comment by kfox1111** [comment]
still an issue?
